### PR TITLE
Future/traversal optimize on vertex and edge

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

Implemented an optimization for Traversals. If you apply a POST filter on
   the vertex and/or edge result this filter will now be applied during the
   traversal to avoid generating the full output for AQL. This will have
   positive effect if you filter on the vertex/edge but return the path,
   this way the system does only need to Produce a path that is allowed to
   be passed through. e.g.

 	  FOR v,e,p IN 10 OUTBOUND @start GRAPH "myGraph"
 		  FILTER v.isRelevant == true
 			RETURN p

 	can now be optimized, and the Traversal statement will only produce
 	paths where the last vertex has `isRelevant == true`.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [x] Backports required for: 3.8

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: js/server/modules/@arangodb/testutils/aql-graph-traversal-generic-tests.js [hunk 1]
```diff
@@ -4517,6 +4518,114 @@ function testCompleteGraphKShortestPathEnabledWeightCheckMultiLimitGen(testGraph
...
+  assertTrue(testGraph.name().startsWith(protoGraphs.completeGraph.name()));
+  const profile = db._query(query, {profile: 2}).getExtra();
+  const result = getCompactStatsNodes(profile).filter(n => n.type === TraversalBlock);
+  // We only have one TraversalBlock
+  assertEqual(result.length, 1);
+  // The traversal block shall only output the numberOfAllowedResults
+  assertEqual(result[0].items, numAllowedResults);
+};
+
+const testCompleteGraphOptimizeVBFS = (testGraph) => assertEarlyFilterOptimization(testGraph,
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply js/server/modules/@arangodb/testutils/aql-graph-traversal-generic-tests.js:4531]: ah .. I see ... Vertex / Edge etc.

**Review by hkernbach — APPROVED** [review]
LGTM. Just minor commentary stuff.

File: arangod/Aql/TraversalConditionFinder.cpp [hunk 2]
```diff
@@ -617,70 +642,83 @@ bool TraversalConditionFinder::before(ExecutionNode* en) {
...
+                conditionToOptimize,
+                [&usedVar, &isSupported](AstNode const* node) -> bool {
+                  if (!IsSupportedNode(usedVar, node)) {
+                    isSupported = false;
+                    return false;
+                  }
+                  return true;
+                },
+                [](AstNode const*) {});
+                */
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/TraversalConditionFinder.cpp:710]: 🤦 jap...

File: arangod/Aql/TraversalNode.cpp [hunk 3]
```diff
@@ -507,6 +517,25 @@ void TraversalNode::toVelocyPackHelper(VPackBuilder& nodes, unsigned flags,
     }
     nodes.close();
   }
+
+  {
+    auto pFilter = postFilterExpression();
+    if (pFilter != nullptr) {
+      nodes.add(VPackValue("postFilter"));
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/TraversalNode.cpp:524]: only this file, and I used same mechanism as for all other values.

File: js/server/modules/@arangodb/testutils/aql-graph-traversal-generic-tests.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mchacki) [reply js/server/modules/@arangodb/testutils/aql-graph-traversal-generic-tests.js:4531]: V and E are variable Names ;)

**Pull request merged**